### ISSUE TITLE
The hash method used for random seed

### ISSUE THREAD

**Issue by souhaibdjaballah** [PacktPublishing/Network-Science-with-Python-and-NetworkX-Quick-Start-Guide#5]
In the code examples, the `hash()` method is used to set the seed. The output is still random and doesn't match the results in the book. I believe this is because the return value of the `hash()` method is _truncated_ according to the host machine's _bit width_ as explained in the official [documentation](https://docs.python.org/3/library/functions.html?highlight=hash#hash).

Regards,

**Comment by salym** [comment]
Hi Souhaib,

You need to set PYTHONHASHSEED to get a kind of reproducibility.

For Jupyter, you can use magic commands:
https://stackoverflow.com/questions/54865930/how-to-set-pythonhashseed-for-jupyter-notebook

For the command line you have 2 ways:
https://keras.io/getting_started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development
http://jimmycallin.com/2016/01/03/make-your-python3-code-reproducible/
By the way, if you are on Win10, in PowerShell you can do it by: $env:PYTHONHASHSEED=0

**Comment by elplatt** [comment]
Good catch @Souhaib-D, and thanks for the info @salym!

This was an issue I noticed when putting the examples together, but didn't have a chance to fix before publication. Because the value of hash() is salted with a seed, as salym points out, it's probably better to use something like hashlib to get a reproducible value. Python random.seed() can take a string directly, but numpy.random.seed() needs an array of integers. Here's an example:


    import random
    from numpy import random as nprand
    import hashlib

    seed = list(map(int, hashlib.sha224(b'Network science rules!').digest()))
    random.seed(seed)
    nprand.seed(seed)

**Comment by Souhaib-D** [comment]
Thank you for your help @salym and @elplatt.

@salym, I tried setting the PYTHONHASHSEED to "0" but it seems this only holds for one runtime (i.e. if you restart the kernel). This has to do with the hash() function algorithm as you mentioned.

@elplatt, you're right the hashlib package is the right option to go about this. Here's my attempt to make it work for both numpy.random and the standard random module.

`import hashlib
seed = int(hashlib.sha224(b"Network Science in Python").hexdigest(), 16) % 2**32`

**Comment by elplatt** [comment]
Glad to hear you've got a working solution. This is valuable feedback, so please feel free to add yourselves to the CONTRIBUTORS.md file. I'll see if there's a way to add this to the book errata. Thanks!

**Comment by Souhaib-D** [comment]
You're welcome @elplatt. And thank you for the great book!
